python/shotgun_data/shotgun_data_retriever.py

```diff
@@ -0,0 +1,723 @@
+        specific.
+
+        :returns:    The Shotgun connection for this instance in the current thread
+        """
+        return self._task_manager.shotgun_connection
+
+    def start(self):
+        """
+        Start the retriever thread
+        """
+        self._task_manager.start_processing()
+
+    def stop(self):
+        """
+        Stop the retriever thread
+        """
+        if not self._task_manager:
+            return
+
+        if self._owns_task_manager:
+            # we own the task manager so we'll need to completely shut it down before
+            # returning
+            self._task_manager.shut_down()
+            self._task_manager = None
+        else:
+            # we don't own the task manager so just stop any tasks we might be running 
+            # and disconnect from it:
+            self._task_manager.stop_task_group(self._bg_tasks_group)
+            self._task_manager.task_completed.disconnect(self._on_task_completed)
+            self._task_manager.task_failed.disconnect(self._on_task_failed)
+            self._task_manager = None
+
+    def clear(self):
+        """
+        Clear the retriever thread job queue
+        """
+        if not self._task_manager:
+            return
+        # stop any tasks running in the task group:
+        self._task_manager.stop_task_group(self._bg_tasks_group)
+
+    def stop_work(self, task_id):
+        """
+        Stop the specified task
+
+        :param task_id: The task to stop
+        """
+        if not self._task_manager:
+            return
+        # stop the task:
+        self._task_manager.stop_task(task_id)
+
+    def get_schema(self, project_id=None):
+        """
+        Execute the schema_read and schema_entity_read methods asynchronously
+
+        :param project_id:  If specified, the schema listing returned will
+                            be constrained by the schema settings for 
+                            the given project.
+        :returns:           A unique task id representing this request.
+        """
+        if not self._task_manager:
+            return
+        task_id = self._task_manager.add_task(self._task_get_schema,
+                                              priority = ShotgunDataRetriever._DOWNLOAD_SCHEMA_PRIORITY,
+                                              group = self._bg_tasks_group,
+                                              project_id = project_id)
+        return str(task_id)
+
+    def execute_find(self, *args, **kwargs):
+        """
+        Executes a Shotgun find query asyncronously.
+
+        This method takes the same parameters as the Shotgun find() call.
+
+        :param *args:       args to be passed to the Shotgun find() call
+        :param **kwargs:    Named parameters to be passed to the Shotgun find() call
+        :returns:           A unique task id representing this request. 
+        """
+        if not self._task_manager:
```

Comment:
This pattern of
if not task_manager -> return
add_task
return str(task_id)
is repeated 7 times. I sure hope there's no missing piece of logic in there or I'd have to copy paste the solution 7 times. ;)
self._add_task?